### PR TITLE
Settings for enable/disable

### LINKED ISSUE #24: Allow enable/disable via settings?
This appears to be the de-facto extension for jupyterlab 3.0, at least until https://github.com/jupyterlab/jupyterlab/pull/9068 lands. Would you be open adding settings-support in repo, so that this end-users can (at least) enable/disable the extension?

I would be happy to implement as a PR if you're open.

Work items:

- [ ] Add settings support via https://github.com/jupyterlab/extension-examples/tree/master/settings. 
- [ ] Add enable/disable flag under settings. Ie. https://github.com/jwkvam/jupyterlab-vim/pull/112
- [ ] Update esc behavior under settings. Ie. https://github.com/ryantam626/jupyterlab-vim/commit/0fd95cf3a3d821f26d438edfcb6a5ec1ae6783e1 and https://github.com/asford/jupyterlab-vim/commit/e02fd37fdbbcb751f381489e37de35f1a500703b

### PR DESCRIPTION
**Pull request by mlucool**
Fixes #24

Please finish the following when submitting a pull request:

- [X] Add a line to `History.md` briefly documenting the change.

If this is a release, additionally do the following:

- [ ] Update the dependencies in `package.json`
- [ ] Run `jlpm install` to update `yarn.lock`

Thanks!

### CODE REVIEW

**Comment by ianhi** [comment]
Hi @mlucool thanks for doing this! (also nice to see you here). Overall I'm :+1:  to this and thanks for taking initiative to clean up the structure a bit.

What do you think of adding an on/off toggle to the `Settings` menu, similar to how jupyterlab git allows for toggling `Simple Staging` via a menu?

**Comment by ianhi** [comment]
Also I think if you add `FEAT` or `feature` to the commit message then the automated version bump machinery will give us a bigger version bump (though @axelfahy would know better than I)

**Comment by mlucool** [comment]
Hi,

This is in the advanced settings menu:
image

If you mean the the top level settings menu, near "Text Editor Keymap", I'm a bit hesitant to until it's built into lab. The git plugin has it's own menu making easy access to changing options cleaner.

**Comment by ianhi** [comment]
> If you mean the the top level settings menu, near "Text Editor Keymap", I'm a bit hesitant to until it's built into lab. The git plugin has it's own menu making easy access to changing options cleaner.

Fair enough. Although, fair warning I was one of the drivers of the "built into lab" and I haven't/won't have time for that for awhile.

**Review by ianhi — CHANGES_REQUESTED** [review]
This is a clear step in the right direction but I think that we shouldn't merge this until we add code to clean up after ourselves when a user disables vim mode. There are currently two issues:

1. Cells that the user has already entered will remain in vim mode
2. All of our commands will remain. So the user won't get back the behavior of `Esc` taking them to the notebook command mode.


I think the fixes for both of these problems can be found in my (now stalled) PR adding vim support to core: https://github.com/jupyterlab/jupyterlab/pull/9068 please feel free to extract them into this PR.

When the settings flip us from `enabled` to `disabled` we need to
1. Use the notebook tracker to loop over all existing cells and ensure that they are no longer in vim mode
2. Save the return statements from when we assign the commands and then call the lumino `dispose` method on them to get rid of them

**Comment by ianhi** [comment]
Expanding a little bit more.

For 1:
Essentially we need to take: https://github.com/jupyterlab/jupyterlab/blob/83a21d729d1c00d87e72674dbbd4f4287fe18f45/packages/codemirror-extension/src/index.ts#L293

For 2:
In that PR I pushed all throwing all the keymap setup into a single function: https://github.com/jupyterlab/jupyterlab/blob/83a21d729d1c00d87e72674dbbd4f4287fe18f45/packages/codemirror-extension/src/index.ts#L205 which would accept the previous keymap commands in order to remove them (this makes it easy to extend to supporting emacs etc) I think it may make sense to do something similar here.

**Comment by mlucool** [comment]
Thanks for the feedback @ianhi. Would you be able to pull those changes on top of this or would you like me to?

**Comment by ianhi** [comment]
> Would you be able to pull those changes on top of this 

I can do this at some point but can't guarantee any timeline as Jlab-vim is mostly a weekends only thing for me right now. (There are just too many interesting things in the world to learn about and to do :disappointed:  )

**Comment by asford** [comment]
Would it be possible to merge this _without_ the "clean disable" flow above as long as the toggle is buried in the settings menu?

I totally agree with @ianhi that a proper cleanup should be done if the setting is exposed as a drop-down in the UI, but a stopgap that requires changing the json settings document then refreshing page seems reasonable-ish in the time being?

**Comment by ianhi** [comment]
> Would it be possible to merge this without the "clean disable" flow above as long as the toggle is buried in the settings menu?

That's a good idea. Prioritize the good over the perfect. I think we should add a comment to the settings file mentioning the need to refresh, but otherwise :+1:  from me.

**Comment by ianhi** [comment]
>  I think we should add a comment to the settings file mentioning the need to refresh, but otherwise +1 from me.

oh I see that's already been done! I'll look into actually merging this tonight or tomorrow

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@axlair/jupyterlab_vim",
-  "version": "0.13.4",
+  "version": "0.13.5-rc.2",
```
Comment (ianhi) [inline package.json:3]: What's the purpose of adding `rc.2`? Won't that make it so that `pip install` doesn't pick up this version?

Comment (mlucool) [reply package.json:3]: We can remove that. I accidently left that during my test of publishing this to an internal registry

**Comment by mlucool** [comment]
> oh I see that's already been done! I'll look into actually merging this tonight or tomorrow

Anything still holding up the merge?

**Comment by ianhi** [comment]
> Anything still holding up the merge?

Apparently only my medium term memory. Sorry about that!

**Review by ianhi — APPROVED** [review]

**Comment by ianhi** [comment]
I'm giving the binder one final test to be sure - but merging and subsequent automated release should be imminent

**Comment by ianhi** [comment]
Thanks @mlucool !

**Comment by ianhi** [comment]
pypi publish action here: https://github.com/axelfahy/jupyterlab-vim/actions/runs/797794893

**Comment by ianhi** [comment]
ack! it failed due to the same error as https://github.com/jupyter-widgets/widget-ts-cookiecutter/issues/101
I think we need to pull in the changes from https://github.com/jupyterlab/extension-cookiecutter-ts/pull/143 to get this to build properly

**Comment by mlucool** [comment]
I ended up needing to do that change a few times already today (which felt very silly to be forced into, although clearly an improvement in code), so one more was somewhat fast: https://github.com/axelfahy/jupyterlab-vim/pull/34

**Pull request merged**